### PR TITLE
Add `uv-workspace` crate with settings discovery and deserialization

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

This PR adds basic struct definitions along with a "workspace" concept for discovering settings. (The "workspace" terminology is used to match Ruff; I did not invent it.)

A few notes:

- We discover any `pyproject.toml` or `uv.toml` file in any parent directory of the current working directory. (We could adjust this to look at the directories of the input files.)
- We don't actually do anything with the configuration yet; but those PRs are large and I want this to be reviewed in isolation.

### CODE REVIEW

File: crates/uv-workspace/src/settings.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+#[serde(deny_unknown_fields, rename_all = "kebab-case")]
+pub struct Options {
+    pub quiet: Option<bool>,
+    pub verbose: Option<bool>,
+    pub native_tls: Option<bool>,
+    pub no_cache: bool,
+    pub cache_dir: Option<PathBuf>,
+    pub pip: Option<PipOptions>,
+    pub resolver: Option<ResolverOptions>,
+    pub installer: Option<InstallerOptions>,
```
Comment (charliermarsh) [inline crates/uv-workspace/src/settings.rs:35]: These should arguably be under `pip`. I'm not totally sure though to be honest.

File: crates/uv/src/main.rs [hunk 2]
```diff
@@ -105,6 +105,9 @@ async fn run() -> Result<ExitStatus> {
         }
     };
 
+    // Load the workspace settings.
+    let _ = uv_workspace::Workspace::find(env::current_dir()?)?;
```
Comment (zanieb) [inline crates/uv/src/main.rs:109]: If we don't have read permissions for the current directory should we error or just not attempt to read a workspace there?

File: crates/uv-workspace/src/workspace.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    ///
+    /// The search starts at the given path and goes up the directory tree until a workspace is
+    /// found.
+    pub fn find(path: impl AsRef<Path>) -> Result<Option<Self>, WorkspaceError> {
+        for ancestor in path.as_ref().ancestors() {
+            if let Some(options) = read_options(ancestor)? {
+                return Ok(Some(Self {
+                    options,
+                    root: ancestor.to_path_buf(),
+                }));
```
Comment (zanieb) [inline crates/uv-workspace/src/workspace.rs:26]: Does this mean that we'll fail on _any_ uv invocation if there's an unparsable `pyproject.toml` file in an ancestor directory?

This behavior kind of drives me crazy in Ruff.

Comment (zanieb) [reply crates/uv-workspace/src/workspace.rs:26]: I'd rather log a warning and return `None`. I think failing on an unparsable `uv.toml` makes more sense to me though.

File: crates/uv-workspace/src/settings.rs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+#[allow(dead_code)]
+#[derive(Debug, Clone, Default, Deserialize)]
+#[serde(deny_unknown_fields, rename_all = "kebab-case")]
+pub struct ResolverOptions {
+    pub resolution: Option<ResolutionMode>,
+    pub prerelease: Option<PreReleaseMode>,
+    pub no_strip_extras: Option<bool>,
+    pub no_annotate: Option<bool>,
+    pub no_header: Option<bool>,
```
Comment (zanieb) [inline crates/uv-workspace/src/settings.rs:66]: Aren't a bunch of these very pip-compile output specific? I find it weird to consider these resolver options.

File: crates/uv-workspace/src/settings.rs (hunk 1 above)
Comment (zanieb) [reply crates/uv-workspace/src/settings.rs:35]: Do we currently have the same `ResolverOptions` and `InstallerOptions` at the top-level _and_ in `pip`?

File: crates/uv-workspace/src/settings.rs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    pub emit_find_links: Option<bool>,
+    pub annotation_style: Option<AnnotationStyle>,
+}
+
+#[allow(dead_code)]
+#[derive(Debug, Clone, Default, Deserialize)]
+#[serde(deny_unknown_fields, rename_all = "kebab-case")]
+pub struct InstallerOptions {
+    pub link_mode: Option<LinkMode>,
+    pub compile: Option<bool>,
```
Comment (zanieb) [inline crates/uv-workspace/src/settings.rs:83]: I wonder if we should choose a clearer name for this

File: crates/uv-workspace/src/settings.rs (hunk 4 above)
Comment (charliermarsh) [reply crates/uv-workspace/src/settings.rs:66]: Yes, absolutely. Hence: https://github.com/astral-sh/uv/pull/3007#discussion_r1566441095

File: crates/uv-workspace/src/workspace.rs (hunk 3 above)
Comment (charliermarsh) [reply crates/uv-workspace/src/workspace.rs:26]: Yes it does.

File: crates/uv-workspace/src/settings.rs (hunk 1 above)
Comment (charliermarsh) [reply crates/uv-workspace/src/settings.rs:35]: No, they're _only_ present at the top-level. Arguably, everything should be put under `pip`. The downside is that it creates a lot of nesting, and we may have options here that _are_ applicable to the non-`pip` interface.

Comment (zanieb) [reply crates/uv-workspace/src/settings.rs:35]: Hm but I see

```
pub struct PipOptions {
    ...
    pub resolver: Option<ResolverOptions>,
    pub installer: Option<InstallerOptions>,
```

Anyway... I'd rather put them in `pip` and pull them into the top-level as we know we need them? I think there are only a couple that would make sense.

Comment (charliermarsh) [reply crates/uv-workspace/src/settings.rs:35]: Maybe I messed that up? Will put them in `pip` and keep them there for now.

File: crates/uv-workspace/src/workspace.rs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+impl Workspace {
+    /// Find the [`Workspace`] for the given path.
+    ///
+    /// The search starts at the given path and goes up the directory tree until a workspace is
+    /// found.
+    pub fn find(path: impl AsRef<Path>) -> Result<Option<Self>, WorkspaceError> {
+        for ancestor in path.as_ref().ancestors() {
+            if let Some(options) = read_options(ancestor)? {
```
Comment (konstin) [inline crates/uv-workspace/src/workspace.rs:22]: Does this mean we only pick the closest file?

The scenario i'm thinking about is one where the work dir is `/home/ferris/foo/bar/` and we have:
* `/home/ferris/uv.toml` or `/home/ferris/.config/uv.toml`: Has some general user-level settings
* `/home/ferris/foo/pyproject.toml`: Defines a workspace of python projects and has some project level settings
* `/home/ferris/foo/bar/pyproject.toml`: A member of that workspace. Does not contain settings (i think?)

In this case, i want to first load user-level setting, then override them with the project level settings (implicitly), then resolve `bar` as part of the `foo` workspace

File: crates/uv-workspace/src/workspace.rs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+/// Represents a project workspace that contains a set of options and a root path.
+#[allow(dead_code)]
+#[derive(Debug, Clone)]
+pub struct Workspace {
+    options: Options,
+    root: PathBuf,
+}
+
+impl Workspace {
```
Comment (konstin) [inline crates/uv-workspace/src/workspace.rs:15]: I'm not sure if i'd call that workspace, i'm thinking about having settings, things that are defined e.g. on the user-level, the project level or the cli level which e.g. define indexes, plus having a workspace, an abstraction over a set of packages that are part of single project. That's the main divergence from ruff, in uv (similar to cargo) you're always in a single workspace and/or a single package; i still have to flesh out how to integrate this into uv.

In cargo, you're technically always in a rust workspace, even if you only define a single crate, that's a single crate workspace. Then there's the concept of [configuration](https://doc.rust-lang.org/cargo/reference/config.html) with hierarchical structure and cli overrides.

**Review by konstin — APPROVED** [review]

File: crates/uv-workspace/src/workspace.rs (hunk 6 above)
Comment (charliermarsh) [reply crates/uv-workspace/src/workspace.rs:22]: Yeah we only pick the closest file. What if we do the same as Ruff, and skip a `pyproject.toml` that lacks a `tool.uv` entry?

File: crates/uv-workspace/src/workspace.rs (hunk 7 above)
Comment (charliermarsh) [reply crates/uv-workspace/src/workspace.rs:15]: I can call this `configuration` instead?

File: crates/uv-workspace/src/workspace.rs (hunk 6 above)
Comment (zanieb) [reply crates/uv-workspace/src/workspace.rs:22]: That'd make sense to me.

File: crates/uv-workspace/src/workspace.rs (hunk 7 above)
Comment (zanieb) [reply crates/uv-workspace/src/workspace.rs:15]: I've been intending to call it a "workspace" whether there's one package or many in your project. I don't know if we need to introduce a separate term?

File: crates/uv-workspace/src/workspace.rs (hunk 6 above)
Comment (konstin) [reply crates/uv-workspace/src/workspace.rs:22]: What do you think about https://doc.rust-lang.org/cargo/reference/config.html#hierarchical-structure ?

File: crates/uv-workspace/src/workspace.rs (hunk 7 above)
Comment (konstin) [reply crates/uv-workspace/src/workspace.rs:15]: I like `configuration`

File: crates/uv-workspace/src/settings.rs (hunk 5 above)
Comment (charliermarsh) [reply crates/uv-workspace/src/settings.rs:83]: `compile_bytecode`? We could change the CLI flag too (and add an alias).

Comment (zanieb) [reply crates/uv-workspace/src/settings.rs:83]: That makes sense to me.

File: crates/uv-workspace/src/workspace.rs (hunk 3 above)
Comment (charliermarsh) [reply crates/uv-workspace/src/workspace.rs]: Fixed this. We show a user-facing warning on `pyproject.toml` but error on `uv.toml`.

**Comment by charliermarsh** [comment]
I'm just gonna leave it as "workspace" for now. Let's revisit when we add _actual_ workspace, that's all internal terminology and abstractions anyway.

**Pull request merged**